### ISSUE TITLE
Provider states without `Parameters` are not  recorded in the contract

### ISSUE THREAD

**Issue by nsfrias** [pact-foundation/pact-go#220]
## Software versions

* **OS: Mac OSX 12.3.1**
* **Consumer Pact library: v2.0.0-beta.11**
* **Provider Pact library: v2.0.0-beta.11**
* **Golang Version: go version go1.18.3 darwin/arm6**
* **Golang environment:**

```
GO111MODULE=""
GOARCH="arm64"
GOBIN=""
GOCACHE="/Users/nunofrias/Library/Caches/go-build"
GOENV="/Users/nunofrias/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="arm64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/nunofrias/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/nunofrias/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/opt/homebrew/Cellar/go/1.18.3/libexec"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/opt/homebrew/Cellar/go/1.18.3/libexec/pkg/tool/darwin_arm64"
GOVCS=""
GOVERSION="go1.18.3"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/nunofrias/Documents/repos/reseller-entitlement-service/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/zn/48yhvjqd17z0x_qy9y6bqn080000gn/T/go-build3642680703=/tmp/go-build -gno-record-gcc-switches -fno-common"
```

## Expected behaviour

When a provider state is set without `Parameters` or the 'Parameters` are an empty map the provider state is not saved in the Pact contract.

Example
```

// No Parameters
Given(ProviderStateV3{Name: "some state",}).
...


// Empty Parameters map
Given(ProviderStateV3{
	Name: "some state",
	Parameters: make(map[string]interface{})
}).
...
```

## Actual behaviour

The state should be recorded in the Pact contract even if no `Parameters` are provided. Not only is this the behaviour supported in other pact languages (namely pact-jvm), but this is a very useful feature as the provider state is a useful way to invoke functions in a specific sequence to set the provider in a specific state. 

One could argue that this functions can be invoked directly on the provider test, however having the provider state explicitly stated on the consumer test makes it evident the steps necessary to have the provider send the expected response / message.

**Comment by mefellows** [comment]
I've just tried to reproduce this with the V3 example in the project but couldn't. 

```func TestConsumerV3(t *testing.T) {
	SetLogLevel("TRACE")

	mockProvider, err := NewV3Pact(MockHTTPProviderConfig{
		Consumer: "PactGoV3Consumer",
		Provider: "V3Provider",
		Host:     "127.0.0.1",
		TLS:      true,
	})
	assert.NoError(t, err)

	// Set up our expected interactions.
	mockProvider.
		AddInteraction().
		Given(ProviderStateV3{
			Name: "User foo exists",
			// Parameters: map[string]interface{}{
			// 	"id": "foo",
			// },
		}).
		UponReceiving("A request to do a foo").
		WithRequest("POST", Regex("/foobar", `\/foo.*`)).
		WithHeader("Content-Type", S("application/json")).
		WithHeader("Authorization", Like("Bearer 1234")).
		WithQuery("baz", Regex("bar", "[a-z]+"), Regex("bat", "[a-z]+"), Regex("baz", "[a-z]+")).
		WithJSONBody(Map{
			"id":       Like(27),
			"name":     FromProviderState("${name}", "billy"),
			"lastName": Like("billy"),
			"datetime": DateTimeGenerated("2020-01-01T08:00:45", "yyyy-MM-dd'T'HH:mm:ss"),
		}).
		WillRespondWith(200).
		WithHeader("Content-Type", S("application/json")).
		WithJSONBody(Map{
			"datetime":       Regex("2020-01-01", "[0-9\\-]+"),
			"name":           S("Billy"),
			"lastName":       S("Sampson"),
			"superstring":    Includes("foo"),
			"id":             Integer(12),
			"accountBalance": Decimal(123.76),
			"itemsMinMax":    ArrayMinMaxLike(27, 3, 5),
			"itemsMin":       ArrayMinLike("thereshouldbe3ofthese", 3),
			"equality":       Equality("a thing"),
			"arrayContaining": ArrayContaining([]interface{}{
				Like("string"),
				Integer(1),
				Map{
					"foo": Like("bar"),
				},
			}),
		})

	// Execute pact test
	err = mockProvider.ExecuteTest(t, test)
	assert.NoError(t, err)
}
```

This test produces the following pact file (with matching rules removed for clarity):

```json
{
  "consumer": {
    "name": "PactGoV3Consumer"
  },
  "interactions": [
    {
      "description": "A request to do a foo",
      "providerStates": [
        {
          "name": "User foo exists"
        }
      ],
      "request": {
        "body": {
          "datetime": "2020-01-01T08:00:45",
          "id": 27,
          "lastName": "billy",
          "name": "billy"
        },
        "generators": {
          "body": {
            "$.datetime": {
              "format": "yyyy-MM-dd'T'HH:mm:ss",
              "type": "DateTime"
            },
            "$.name": {
              "expression": "${name}",
              "type": "ProviderState"
            }
          }
        },
        "headers": {
          "Authorization": "Bearer 1234",
          "Content-Type": "application/json"
        },

        "method": "POST",
        "path": "/foobar",
        "query": {
          "baz": [
            "bar",
            "bat",
            "baz"
          ]
        }
      },
      "response": {
        "body": {
          "accountBalance": 123.76,
          "arrayContaining": [
            "string",
            1,
            {
              "foo": "bar"
            }
          ],
          "datetime": "2020-01-01",
          "equality": "a thing",
          "id": 12,
          "itemsMin": [
            "thereshouldbe3ofthese",
            "thereshouldbe3ofthese",
            "thereshouldbe3ofthese"
          ],
          "itemsMinMax": [
            27,
            27,
            27,
            27,
            27
          ],
          "lastName": "Sampson",
          "name": "Billy",
          "superstring": "foo"
        },
        "headers": {
          "Content-Type": "application/json"
        },
        "status": 200
      }
    }
  ],
  "metadata": {
    "pactRust": {
      "ffi": "0.3.2",
      "mockserver": "0.9.2",
      "models": "0.4.1"
    },
    "pactSpecification": {
      "version": "3.0.0"
    }
  },
  "provider": {
    "name": "V3Provider"
  }
}
```

**Comment by nsfrias** [comment]
Hi Matt.

That is precisely the case. Sorry for missing that fact in the issue report.

**Comment by nsfrias** [comment]
This has been fixed in v2.0.0-beta.14 of pact-go.

Thank you.
